thunder/core/codeutils.py

```diff
@@ -121,8 +121,21 @@ def to_printable(
         return x
 
     if is_collection(x):
+        if type(x) not in (
+            dict,
+            list,
+            str,
+            int,
+            bool,
+            tuple,
+            torch.dtype,
+            float,
+            dtypes.floating,
+            devices.Device,
+            torch.memory_format,
+        ):
```

Comment:
What about using [`isinstance(x, ...)`](https://docs.python.org/3.10/library/functions.html#isinstance) instead? With that you could check if the type is or inherits from a collection from [`collections.abc`](https://docs.python.org/3.10/library/collections.abc.html#module-collections.abc) 👀

Reply:
Do you mean something along the lines of `isinstance(x, collections.abc.Container)`? If that is the case, from my understanding, this will be `True` for HF ModelOutput (`class 'transformers.modeling_outputs.BaseModelOutput'`) which is not what we want. Let me know if I misunderstood anything! :)

Reply:
As @k223kim explains, we tried that, but so HF ModelOutput inherits OrderedDict (and thus dict), but cannot be handled.

Reply:
Oh I see, why did we choose to go for a whitelist logic instead of blacklisting the `OrderedDict`? 
i.e. : 
``` python
if isinstance(args, OrderedDict):
    raise TypeError(f"tree_flatten of type {type(args)} is not supported.")
```

Reply:
So the idea is to reliably avoid the recursion error and present a reasonable message instead. tree_flatten has a positive list of things it works with, so that's what we did here. My only hesitation is the interaction with @kshitij12345 's proposed patch for #632, which I want to figure out today-ish.